### DISCUSSION TITLE
Multiple devices connection

### DISCUSSION THREAD

**Discussion by rasbianer** [Jakeler/ble-serial#52] [Q&A] [answered]
Does the library support opening different virtual serial ports at the same time?
I am trying to connect  a Raspberry Pi 4 to several Arduinos and their BLE modules.
Is there a special command to run from the command prompt? Is there any option I should enable while calling the command I normally use, which is: $ ble-serial -d <baddr>
I tried to use 'ble_serial' as a function to run on Python's multiple threads, but I couldn't figure out how to call the function and how to give to it the device's address I want to connect to and the port's number I want to choose.
Thanks in advance for any help.

**Comment by Jakeler** [comment, marked as answer]
Yes, this works. Just start it multiple times, in separate processes.
I would not recommend to import it in as python module, while it is certainly possible to pass the correct parameters, ble-serial is not really designed for it. Use the command line interface instead. 
To run it in parallel you can use the shell background job feature. Add the `-p` (port) option to make sure every instance has a unique path. Also you probably want to redirect the log output to keep it separate. Resulting command lines could look like:
```terminal
$ ble-serial -d $ADDR1 -p /tmp/ttyBLE1 2> dev1.log &
[1] 178378

$ ble-serial -d $ADDR2 -p /tmp/ttyBLE2 2> dev2.log &
[2] 178397
```
The `&`  at the end causes it to go into background mode immediately, showing the job nr and PID. 
Now it is usual control, here I can see 2 running with `jobs` as expected:
```
$ jobs                                                     
[1]  - running   ble-serial -d ... -p /tmp/ttyBLE1 2> dev1.log &
[2]  + running   ble-serial -d ... -p /tmp/ttyBLE2 2> dev2.log &
```
Get them back to foreground with `fg` or view logs live with `tail -f dev1.log` etc.
Of course you can also put everything into a shell script, for stopping multiple jobs you can send signals with  `kill -s SIGINT $PID` instead of going through them manually with `fg`. This will bring ble-serial to graceful shutdown as well, same as ctrl-c.

**Reply by rasbianer** [reply]
thank you, this was super helpful and it worked on the command prompt as you suggested.
Thank you so much again
